File: chrome/android/features/autofill_assistant/java/src/org/chromium/chrome/browser/autofill_assistant/AssistantBottomBarDelegate.java
// Copyright 2020 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.browser.autofill_assistant;

/** Common interface for the bottom bar delegate. */
public interface AssistantBottomBarDelegate {
    /** The back button has been pressed. */
    boolean onBackButtonPressed();

    // TODO(micantox): Instead of a dedicated notification, we should just notify the controller of
    // the new bottom sheet state and have the logic to shutdown there. Currently, this would be
    // tricky to do because it would interfere with the existing Controller::SetBottomSheetState
    // method and in particular tab switching.
    /**
     * The bottom sheet was closed with a swipe gesture. Note that this will be fired both when
     * going into the PEEK state as well as when dismissing the sheet altogether.
     */
    void onBottomSheetClosedWithSwipe();
}

File: chrome/browser/android/autofill_assistant/assistant_bottom_bar_delegate.cc
// Copyright 2020 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "chrome/browser/android/autofill_assistant/assistant_bottom_bar_delegate.h"

#include "chrome/android/features/autofill_assistant/jni_headers/AssistantBottomBarNativeDelegate_jni.h"
#include "chrome/browser/android/autofill_assistant/ui_controller_android.h"

using base::android::AttachCurrentThread;

namespace autofill_assistant {

AssistantBottomBarDelegate::AssistantBottomBarDelegate(
    UiControllerAndroid* ui_controller)
    : ui_controller_(ui_controller) {
  java_assistant_bottom_bar_delegate_ =
      Java_AssistantBottomBarNativeDelegate_create(
          AttachCurrentThread(), reinterpret_cast<intptr_t>(this));
}

AssistantBottomBarDelegate::~AssistantBottomBarDelegate() {
  Java_AssistantBottomBarNativeDelegate_clearNativePtr(
      AttachCurrentThread(), java_assistant_bottom_bar_delegate_);
}

bool AssistantBottomBarDelegate::OnBackButtonClicked(
    JNIEnv* env,
    const base::android::JavaParamRef<jobject>& jcaller) {
  return ui_controller_->OnBackButtonClicked();
}

void AssistantBottomBarDelegate::OnBottomSheetClosedWithSwipe(
    JNIEnv* env,
    const base::android::JavaParamRef<jobject>& jcaller) {
  ui_controller_->OnBottomSheetClosedWithSwipe();
}

base::android::ScopedJavaGlobalRef<jobject>
AssistantBottomBarDelegate::GetJavaObject() {
  return java_assistant_bottom_bar_delegate_;
}

}  // namespace autofill_assistant
